Source/Core/DolphinQt2/Config/LoggerWidget.cpp

```diff
@@ -0,0 +1,489 @@
+      {
+        LogManager::GetInstance()->RemoveListener(log_type, LogListener::LOG_WINDOW_LISTENER);
+      }
+    }
+  }
+
+  m_ini_options->Set("WriteToFile", m_out_file->isChecked());
+  m_ini_options->Set("WriteToConsole", m_out_console->isChecked());
+  m_ini_options->Set("WriteToWindow", m_out_window->isChecked());
+
+  // Config - Log Types
+  for (int i = 0; i < LogTypes::NUMBER_OF_LOGS; ++i)
+  {
+    const auto type = static_cast<LogTypes::LOG_TYPE>(i);
+    const auto name = LogManager::GetInstance()->GetShortName(type);
+    bool enabled = m_types_list->item(i)->checkState() == Qt::Checked;
+    bool was_enabled = LogManager::GetInstance()->IsEnabled(type);
+
+    if (enabled != was_enabled)
+    {
+      LogManager::GetInstance()->SetEnable(type, enabled);
+      m_ini_logs->Set(name, enabled);
+    }
+  }
+
+  // Log - Wrap Lines
+  m_ini_options->Set("WrapLines", m_log_wrap->isChecked());
+  m_log_text->setLineWrapMode(m_log_wrap->isChecked() ? QTextEdit::WidgetWidth : QTextEdit::NoWrap);
+
+  // Log - Font Selection
+  m_ini_options->Set("Font", m_log_font->currentIndex());
+  UpdateFont();
+
+  m_ini.Save(File::GetUserPath(F_LOGGERCONFIG_IDX));
+}
+
+void LoggerWidget::OnTabVisibilityChanged()
+{
+  bool log_visible = Settings::Instance().IsLogVisible();
+  bool config_visible = Settings::Instance().IsLogConfigVisible();
+
+  m_tab_widget->clear();
+
+  if (log_visible)
+    m_tab_widget->addTab(m_tab_log, tr("Log"));
+
+  if (config_visible)
+    m_tab_widget->addTab(m_tab_config, tr("Configuration"));
+
+  if (!log_visible && !config_visible)
+    hide();
+  else
+    show();
+}
+
+void LoggerWidget::Log(LogTypes::LOG_LEVELS level, const char* text)
+{
+  std::lock_guard<std::mutex> lock(m_log_mutex);
+
+  const char* color = "white";
+
+  switch (level)
+  {
+  case LogTypes::LOG_LEVELS::LERROR:
+    color = "red";
+    break;
+  case LogTypes::LOG_LEVELS::LWARNING:
+    color = "yellow";
+    break;
+  case LogTypes::LOG_LEVELS::LNOTICE:
+    color = "green";
+    break;
+  case LogTypes::LOG_LEVELS::LINFO:
+    color = "cyan";
+    break;
+  case LogTypes::LOG_LEVELS::LDEBUG:
+    color = "lightgrey";
+    break;
+  }
+
+  m_log_queue.push(QStringLiteral("%1 <font color='%2'>%3</font>")
+                       .arg(QString::fromStdString(std::string(text).substr(0, TIMESTAMP_LENGTH)),
+                            QString::fromStdString(color),
+                            QString::fromStdString(std::string(text).substr(TIMESTAMP_LENGTH))));
+}
+
+bool LoggerWidget::eventFilter(QObject* object, QEvent* event)
+{
+  if (event->type() == QEvent::Hide || event->type() == QEvent::Resize ||
+      event->type() == QEvent::Move)
+  {
+    Settings::Instance().setValue(QStringLiteral("LoggerWidget/geometry"), saveGeometry());
+    Settings::Instance().setValue(QStringLiteral("LoggerWidget/floating"), isFloating());
+  }
+
+  if (event->type() == QEvent::Close)
+  {
+    if (Settings::Instance().IsLogVisible() != false ||
```

Comment:
Just thought it was neater, if it seems to be the common opinion I'm fine with changing it. (It's simpler anyway.)